### PR TITLE
Remove windows specific juliarc.jl

### LINKED ISSUE #18769: Method definition warning that only shows up on Windows
On windows

``` julia
> julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.0 (2016-09-19 18:14 UTC)
 _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org/ release
|__/                   |  x86_64-w64-mingw32

julia> *{T}(x::Int, ::Type{T}) = T(x)
ERROR: error in method definition: function Base.* must be explicitly imported to be extended
```

However on Linux I don't see this warning

### PR DESCRIPTION
**Pull request by musm**
close https://github.com/JuliaLang/julia/issues/18769

### CODE REVIEW

**Comment by musm** [comment]
embedding example is failing for some reason, any ideas?

**Comment by musm** [comment]
@tkelman  anyideas on the embedding failure, is busybox or 7z somehow needed for that ? Not sure how to test locally last time I tried to build on windows using the readme was not successful .

**Comment by tkelman** [comment]
> last time I tried to build on windows using the readme was not successful .

You're going to need to be more specific than "not successful"

**Comment by tkelman** [comment]
For the embedding executable to run on windows, libjulia either needs to be copied into the same directory as the embedding executable, or we can do the path appending as part of `embedding-test.jl`

File: examples/embedding/embedding-test.jl [hunk 1]
```diff
@@ -3,6 +3,11 @@
 # tests the output of the embedding example is correct
 using Base.Test
 
+if is_windows()
+    # libjulia needs to be in the same directory as the embedding executable
```
Comment (tkelman) [inline examples/embedding/embedding-test.jl]: , or on the path

Comment (musm) [inline examples/embedding/embedding-test.jl]: updated

Comment (musm) [inline examples/embedding/embedding-test.jl]: Feel free to update the makefile to also copy libjulia if you feel that it's also useful to do that for external users copying the makefile

**Comment by musm** [comment]
great tests now pass

**Comment by musm** [comment]
not a lot of packages that explicitly use 7z

**Comment by musm** [comment]
bump, FYI this will need a tag on WinRPM and BinDeps before merging

**Comment by musm** [comment]
This is g2g on my end just waiting on BinDeps and WinRPM getting tagged...

**Comment by musm** [comment]
WinRPM and BinDeps have been tagged, this should be good to go now. Unless someone requests something else.

**Comment by musm** [comment]
@tkelman I hate to keep bumping, but this is ready from my end. If there is a good reason to hold off for a bit it's not a problem, just let me know.

FYI the only remaining tagged package that has not merged my PR for this 7z change is `Vectorize` which hasn't has an update in 8 months, the other two are not registered.
Thanks.

**Comment by tkelman** [comment]
The problems this is likely to cause are bigger than any it fixes. We can find out what those are going to be when more people are testing against nightly separately from worrying about making their packages work on 0.6, but windows-specific problems require users or appveyor testing to identify so it takes some time. This isn't solving an important enough problem to be worth diverting anyone's attention away from 0.6 at the moment.

**Comment by musm** [comment]
Thanks for the clarification, sounds reasonable. This can wait until more firmly within the 0.7 timeframe.

**Comment by musm** [comment]
Any chance this could now be considered? There shouldn't any significant user facing changes with registered packages  since I  have covered them with PRs.

**Comment by musm** [comment]
bump

**Comment by musm** [comment]
bump, anything preventing this?

**Comment by musm** [comment]
bump

**Comment by musm** [comment]
bump

**Comment by tkelman** [comment]
This should be mentioned in NEWS actually

**Comment by musm** [comment]
Updated to include NEWS entry.

File: NEWS.md [hunk 2]
```diff
@@ -175,6 +175,8 @@ Deprecated or removed
   * The forms of `read`, `readstring`, and `eachline` that accepted both a `Cmd` object and an
     input stream are deprecated. Use e.g. `read(pipeline(stdin, cmd))` instead ([#22762]).
 
+  * The default `juliarc.jl` file on Windows has been removed. Use `joinpath(JULIA_HOME, "7z.exe")`
+    if you need access, e.g., to `7z.exe` on Windows ([#21540]).
```
Comment (tkelman) [inline NEWS.md]: there's nothing specific to 7z about this - it's the case for any executable or library in julia's bin dir

Comment (musm) [inline NEWS.md]: Correct, that is why I put an `e.g.` in the entry. Will update since it may not be clear.

File: NEWS.md [hunk 3]
```diff
@@ -175,6 +175,9 @@ Deprecated or removed
   * The forms of `read`, `readstring`, and `eachline` that accepted both a `Cmd` object and an
     input stream are deprecated. Use e.g. `read(pipeline(stdin, cmd))` instead ([#22762]).
 
+  * The default `juliarc.jl` file on Windows has been removed. Now must explicitly include the
+    file if you need access to executables or libraries in the `JULIA_HOME` directory, e.g.
```
Comment (tkelman) [inline NEWS.md]: the full path

**Comment by iblis17** [comment]
The FreeBSD CI build error is caused by #22566

**Comment by musm** [comment]
Interesting (unrelated) failure on 32 bit windows
```julia
linalg/schur (4)          |   28.12  |  0.27  |  1.0 | 731.36     | 216.48  
	From worker 4:	
	From worker 4:	This application has requested the Runtime to terminate it in an unusual way.
	From worker 4:	Please contact the application's support team for more information.
	From worker 4:	
	From worker 4:	signal (22): SIGABRT
	From worker 4:	while loading C:\projects\julia\julia-9e219c0614\share\julia\test\linalg/special.jl, in expression starting on line 172
	From worker 4:	crt_sig_handler at C:/projects/julia/src\signals-win.c:92
	From worker 4:	raise at C:\windows\system32\msvcrt.dll (unknown line)
	From worker 4:	abort at C:\windows\system32\msvcrt.dll (unknown line)
	From worker 4:	.text at C:\projects\julia\julia-9e219c0614\bin\libjulia.dll (unknown line)
	From worker 4:	_assert at C:\projects\julia\julia-9e219c0614\bin\libjulia.dll (unknown line)Worker 4 terminated.
	From worker 4:	jl_typemap_level_assoc_exact at C:/projects/julia/src\typemap.c:843ERROR (unhandled task failure): read: connection reset by peer (ECONNRESET)
	From worker 4:	jl_typemap_assoc_exact at C:/projects/julia/src\julia_internal.h:925 [inlined]
	From worker 4:	jl_lookup_generic_ at C:/projects/julia/src\gf.c:1872 [inlined]
	From worker 4:	jl_apply_generic at C:/projects/julia/src\gf.c:1924
	From worker 4:	vcat at .\sparse\sparsematrix.jl:3067
	From worker 4:	jl_call_fptr_internal at C:/projects/julia/src\julia_internal.h:369 [inlined]
	From worker 4:	jl_call_method_internal at C:/projects/julia/src\julia_internal.h:388 [inlined]
	From worker 4:	jl_apply_generic at C:/projects/julia/src\gf.c:1927
	From worker 4:	jl_apply at C:/projects/julia/src\julia.h:1447 [inlined]
	From worker 4:	jl_f__apply at C:/projects/julia/src\builtins.c:422
	From worker 4:	vcat at .\sparse\sparsevector.jl:989
	From worker 4:	jl_call_fptr_internal at C:/projects/julia/src\julia_internal.h:369 [inlined]
	From worker 4:	jl_call_method_internal at C:/projects/julia/src\julia_internal.h:388 [inlined]
	From worker 4:	jl_apply_generic at C:/projects/julia/src\gf.c:1927
	From worker 4:	macro expansion at C:\projects\julia\julia-9e219c0614\share\julia\test\linalg\special.jl:223 [inlined]
	From worker 4:	anonymous at .\<missing> (unknown line)
	From worker 4:	jl_call_fptr_internal at C:/projects/julia/src\julia_internal.h:369 [inlined]
	From worker 4:	jl_call_method_internal at C:/projects/julia/src\julia_internal.h:388 [inlined]
	From worker 4:	jl_toplevel_eval_flex at C:/projects/julia/src\toplevel.c:594
	From worker 4:	jl_parse_eval_all at C:/projects/julia/src\ast.c:908
	From worker 4:	include_string at .\loading.jl:520
	From worker 4:	jlcall_include_string_33830 at C:\projects\julia\julia-9e219c0614\lib\julia\sys.dll (unknown line)
	From worker 4:	jl_call_fptr_internal at C:/projects/julia/src\julia_internal.h:369 [inlined]
	From worker 4:	jl_call_method_internal at C:/projects/julia/src\julia_internal.h:388 [inlined]
	From worker 4:	jl_apply_generic at C:/projects/julia/src\gf.c:1927
	From worker 4:	include_from_node1 at .\loading.jl:559
	From worker 4:	jlcall_include_from_node1_18732 at C:\projects\julia\julia-9e219c0614\lib\julia\sys.dll (unknown line)
	From worker 4:	jl_call_fptr_internal at C:/projects/julia/src\julia_internal.h:369 [inlined]
	From worker 4:	jl_call_method_internal at C:/projects/julia/src\julia_internal.h:388 [inlined]
	From worker 4:	jl_apply_generic at C:/projects/julia/src\gf.c:1927
	From worker 4:	include at .\sysimg.jl:14 [inlined]
	From worker 4:	include at C:\projects\julia\julia-9e219c0614\share\julia\test\testdefs.jl:11
	From worker 4:	macro expansion at C:\projects\julia\julia-9e219c0614\share\julia\test\testdefs.jl:18 [inlined]
	From worker 4:	macro expansion at .\test.jl:943 [inlined]
	From worker 4:	macro expansion at .\util.jl:379 [inlined]
	From worker 4:	macro expansion at C:\projects\julia\julia-9e219c0614\share\julia\test\testdefs.jl:17 [inlined]
	From worker 4:	anonymous at .\<missing> (unknown line)
	From worker 4:	jl_call_fptr_internal at C:/projects/julia/src\julia_internal.h:369 [inlined]
	From worker 4:	jl_call_method_internal at C:/projects/julia/src\julia_internal.h:388 [inlined]
	From worker 4:	jl_toplevel_eval_flex at C:/projects/julia/src\toplevel.c:594 [inlined]
	From worker 4:	jl_toplevel_eval at C:/projects/julia/src\toplevel.c:605
	From worker 4:	jl_toplevel_eval_in at C:/projects/julia/src\builtins.c:492
	From worker 4:	eval at .\boot.jl:248
	From worker 4:	jlcall_eval_27128 at C:\projects\julia\julia-9e219c0614\lib\julia\sys.dll (unknown line)
	From worker 4:	jl_call_fptr_internal at C:/projects/julia/src\julia_internal.h:369 [inlined]
	From worker 4:	jl_call_method_internal at C:/projects/julia/src\julia_internal.h:388 [inlined]
	From worker 4:	jl_apply_generic at C:/projects/julia/src\gf.c:1927
	From worker 4:	runtests at C:\projects\julia\julia-9e219c0614\share\julia\test\testdefs.jl:21
	From worker 4:	runtests at C:\projects\julia\julia-9e219c0614\share\julia\test\testdefs.jl:4
	From worker 4:	unknown function (ip: 0000000014208566)
	From worker 4:	jl_call_fptr_internal at C:/projects/julia/src\julia_internal.h:369 [inlined]
	From worker 4:	jl_call_method_internal at C:/projects/julia/src\julia_internal.h:388 [inlined]
	From worker 4:	jl_apply_generic at C:/projects/julia/src\gf.c:1927
	From worker 4:	jl_apply at C:/projects/julia/src\julia.h:1447 [inlined]
	From worker 4:	jl_f__apply at C:/projects/julia/src\builtins.c:422
	From worker 4:	#106 at .\distributed\process_messages.jl:268
	From worker 4:	run_work_thunk at .\distributed\process_messages.jl:56
	From worker 4:	macro expansion at .\distributed\process_messages.jl:268 [inlined]
	From worker 4:	#105 at .\event.jl:73
	From worker 4:	jl_call_fptr_internal at C:/projects/julia/src\julia_internal.h:369 [inlined]
	From worker 4:	jl_call_method_internal at C:/projects/julia/src\julia_internal.h:388 [inlined]
	From worker 4:	jl_apply_generic at C:/projects/julia/src\gf.c:1927
	From worker 4:	jl_apply at C:/projects/julia/src\julia.h:1447 [inlined]
	From worker 4:	start_task at C:/projects/julia/src\task.c:267
	From worker 4:	Allocations: 25880915 (Pool: 25878365; Big: 2550); GC: 61
	From worker 4:	Assertion failed!
	From worker 4:	
	From worker 4:	Program: C:\projects\julia\julia-9e219c0614\bin\julia.exe
	From worker 4:	File: C:/projects/julia/src/typemap.c, Line 843
	From worker 4:	
	From worker 4:	Expression: jl_is_datatype(ty)
	From worker 4:	
	From worker 3:	read: connection reset by peer (ECONNRESET)CapturedException(read: connection reset by peer (ECONNRESET)
linalg/triangular (3)     |  367.64  |  6.70  |  1.8 | 9889.59    | 655.69  
	From worker 1:		From worker 1:	INFO: SO_REUSEPORT is unsupported, skipping reuseport tests.
linalg/special: Error During Test
  Test threw an exception of type ProcessExitedException
  Expression: linalg/special
  ProcessExitedException()
linalg/eigen: Error During Test
  Test threw an exception of type ProcessExitedException
  Expression: linalg/eigen
  ProcessExitedException()
linalg/bunchkaufman: Error During Test
  Test threw an exception of type ProcessExitedException
```

**Comment by musm** [comment]
The error looks similar too https://github.com/JuliaLang/julia/issues/11083

**Comment by musm** [comment]
restarting for travis

**Comment by musm** [comment]
Unrelated errors on Travis. Is this good to go now?

**Comment by musm** [comment]
@tkelman can we proceed with this? It would be nice to have the same experience as in linux and since the juliarc file is not needed it makes sense to me to remove it. FWIW I hit the original issue twice (first time to report the issue, which was noted by `totalverb`, second time I opened this PR to fix the issue) Thanks.

**Comment by musm** [comment]
Rebased PR.

**Comment by musm** [comment]
bump

**Comment by tkelman** [comment]
I don't personally think the issue this fixes is consequential, but the changes it's making could be. Windows shared libraries are searched for in PATH, on other platforms that's handled differently.

**Comment by musm** [comment]
> I don't personally think the issue this fixes is consequential, but the changes it's making could be. Windows shared libraries are searched for in PATH, on other platforms that's handled differently


I think all the libs are loaded when julia is started irrespective of whether there are in the .juliarc file. For example  I haven't ran into any issues with running julia without this file  and calling functions that are linked to the lib in the `bin` folder, such as `eig` which properly seems to find `openlibblas` without the addition to the juliarc file: (see below)
image


@vtjnash  you made this comment here https://github.com/JuliaLang/julia/issues/18769#issuecomment-296852629 
do you have any thoughts on this?


I agree it's not necessary, but I hate stuff that exists or is there without a purpose. I just would like to confirm that this is in fact needed before closing the PR.

**Comment by tkelman** [comment]
It's not pure Julia code I'm concerned about, it's packages or external programs that might be trying to find libraries included in Julia, and may have a harder time doing so with this change.

**Comment by musm** [comment]
How about we test this out and if there are any reported issues, we can revert the change?

**Comment by StefanKarpinski** [comment]
The trouble is that this won't get widespread testing until after it's too late to revert this (i.e. 1.0). Still, I guess we're going ahead with trying it.

**Comment by musm** [comment]
Fair enough, feel free to revert .

**Comment by tkelman** [comment]
If it is determined to be a problem during 0.7 feature freeze or rc's, I don't think it would break anything to revert even fairly late in the 0.7 release cycle.

**Pull request merged**